File: expansion_board/i2c/mpu6050/python/main.py
###############################################
#
#  file: mpu6050.py
#  update: 2024-10-22
#  usage: 
#      sudo python mpu6050.py
#
###############################################

import time
import gpiod
import os
import struct
import subprocess
import math
import board
import busio
from PIL import Image, ImageDraw, ImageFont
import adafruit_ssd1306
import threading
import evdev
from mpu6050 import MPU6050

# oled初始化
oled_init_flag = 1
try:
    oled_i2c = busio.I2C(board.I2C3_SCL, board.I2C3_SDA)
    disp = adafruit_ssd1306.SSD1306_I2C(128, 64, oled_i2c)

    width = disp.width  
    height = disp.height  
    image = Image.new('1', (width, height))  
    draw = ImageDraw.Draw(image)
    draw.rectangle((0, 0, width, height), outline=0, fill=0)

    font = ImageFont.truetype("NotoSerifCJKhk-VF.ttf", 12, encoding="unic")  
    x = 5   
    top = 5  
except Exception as e:
    oled_init_flag = 0
#-----------------------------------------------------------------

# MPU6050初始化
mpu6050_i2c = busio.I2C(board.I2C5_SCL, board.I2C5_SDA)
mpu6050 = MPU6050(mpu6050_i2c)
mpu6050.start()
###################################################################

def main():

    try:

        while(1):
            
            accelX, accelY, accelZ = mpu6050.readAcc()      # 读取加速度数据
            gyroX, gyroY, gyroZ = mpu6050.readGyro()        # 读取陀螺仪数据

            accelXStr = str("%.2f"%accelX)                  
            accelYStr = str("%.2f"%accelY)
            accelZStr = str("%.2f"%accelZ)

            gyroXStr = str("%.2f"%gyroX)
            gyroYStr = str("%.2f"%gyroY)
            gyroZStr = str("%.2f"%gyroZ)

            if oled_init_flag != 0:
                draw.rectangle((1, 1, width-1, height-1), outline=1, fill=0)

                # 绘制加速度数据文本
                draw.text((5, 0), "accel", font=font, fill=255)
                draw.text((5, 16), accelXStr, font=font, fill=255)
                draw.text((5, 32), accelYStr, font=font, fill=255)
                draw.text((5, 45), accelZStr, font=font, fill=255)

                # 绘制陀螺仪数据文本
                draw.text((40, 0), "gyro", font=font, fill=255)
                draw.text((40, 16), gyroXStr, font=font, fill=255)
                draw.text((40, 32), gyroYStr, font=font, fill=255)
                draw.text((40, 45), gyroZStr, font=font, fill=255)

                # 刷新显示
                disp.image(image)  
                disp.show() 

            print("accel:", "x:", accelXStr, "y:", accelYStr, "z:", accelZStr, "m/s^2")
            print("gyro:", "x:", gyroXStr, "y:", gyroYStr, "z:", gyroZStr, "deg/s")
            print("\n")

            time.sleep(0.3)

    except Exception as e:
        
        print("exit...：", e)
        
    finally:
    
        mpu6050.close()

        if oled_init_flag != 0:
            # oled清屏
            disp.fill(0)
            disp.show() 

if __name__ == "__main__":  
    main()

File: expansion_board/i2c/mpu6050/c/main.c
/*
*
*   file: main.c
*   update: 2024-10-23
*   usage: 
*       make
*       sudo ./main
*
*/

#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <stdio.h>
#include <string.h>
#include <sys/ioctl.h>
#include <poll.h>
#include <stdint.h>
#include <signal.h> 
#include <linux/input.h>
#include <math.h>
#include <pthread.h>
#include <time.h>

#include "oled.h"
#include "mpu6050.h"

int oled_init_flag = 1;

void sigint_handler(int sig_num) 
{    
    /* oled清屏 */
    if(oled_init_flag != 0)
        oled_clear();

    mpu6050_exit();

    exit(0);  
}

int main(int argc, char **argv)
{
    int ret = 0;

    short int accel_buf[3];
    short int gyro_buf[3];
    
    float lsb_sensitivity_accel = 8192;
    float lsb_sensitivity_gyro = 16.4;

    char accel_x_str[20] = {0};
    char accel_y_str[20] = {0};
    char accel_z_str[20] = {0};

    char gyro_x_str[20] = {0};
    char gyro_y_str[20] = {0};
    char gyro_z_str[20] = {0};

    /* register exit signal ( Ctrl + c ) */
    signal(SIGINT, sigint_handler);

    /* oled初始化 */
    ret = oled_init(3);
    if(ret == -1)
    {
        printf("oled init err!\n");
        oled_init_flag = 0;
    }

    /* mpu6050初始化 */
    ret = mpu6050_init(5);
    if(ret == -1)
    {
        printf("mpu6050 init err!\n");
        return -1;
    }

    while(1)
    {   
        mpu6050_read_gyro(gyro_buf);
        mpu6050_read_accel(accel_buf);
        
        sprintf(accel_x_str, "%.2f0", accel_buf[0]/lsb_sensitivity_accel);
        sprintf(accel_y_str, "%.2f0", accel_buf[1]/lsb_sensitivity_accel);
        sprintf(accel_z_str, "%.2f0", accel_buf[2]/lsb_sensitivity_accel);
        
        sprintf(gyro_x_str, "%.2f0", gyro_buf[0]/lsb_sensitivity_gyro);
        sprintf(gyro_y_str, "%.2f0", gyro_buf[1]/lsb_sensitivity_gyro);
        sprintf(gyro_z_str, "%.2f0", gyro_buf[2]/lsb_sensitivity_gyro);

        if(oled_init_flag != 0)
        {
            oled_show_string(2, 0, "accel");
            oled_show_string(2, 2, accel_x_str);
            oled_show_string(2, 4, accel_y_str);
            oled_show_string(2, 6, accel_z_str);

            oled_show_string(60, 0, "gyro");
            oled_show_string(60, 2, gyro_x_str);
            oled_show_string(60, 4, gyro_y_str);
            oled_show_string(60, 6, gyro_z_str);
        }

        printf("accel: x: %s, y: %s, z: %s (m/s^2)\n", accel_x_str, accel_y_str, accel_z_str);
        printf("gyro: x: %s, y: %s, z: %s (deg/s)\n", gyro_x_str, gyro_y_str, gyro_z_str);
        printf("\n");

        sleep(0.3);
    }

    return 0;
}